### PR TITLE
Max Item Size for Storages

### PR DESCRIPTION
**Pull request by xTheLifex**
## About the PR 

Implemented a "Max Item Size" for storages; allowing to balance out some storages where we might want a large amount of only small items, without using a **whitelist**.
Bypassing an existing whitelist with this is possible as it also includes a bool for **whitelist bypass**.
Should not affect the current storages until the storage items get a "maxItemSize" value set in the YML.

This should allow for more balancing options on storages. 

**Example usage:**
A hypothetical purse storage item holds 30 of space. A mop takes 10 of space. It would be possible to fit a mop inside a purse, which is unrealistic and counterintuitive. For this we might come to a few solutions in mind:

- A whitelist, but then, only allowed items would be let in the purse. And as more and more items get added, maintaining such whitelists can become tiring and unefficient, as there could potentially be hundreds of  small sized items that _should_ fit inside a purse.
- Lowering the purse's storage capacity. Sure! But now we can only fit like 5 coins inside a purse... (Assuming coins would t ake up 1 in space)

Then, what do we do? Well. I propose a maximum item size for storages. In addition to checking if the item can fit the capacity and does not surpass it, we can also check if it's below the storage's Max Item Size.

Back to our purse. Our purse has a capacity of 30, but a max item size of 5. This means that even though a mop's size is 10, it's above the purse's maximum size for an item to be put inside it. So it will not fit. And we can fit 30 coins inside (and probably more with changes to those variables).
In theory, you could have a coin purse with 100 space but a max item size of 1, so only a coin or a cigarrete will fit inside it

In short, this is just another suggestion on a new way to balance certain storages.

**Changelog**

:cl: TheLife
- add: Added a maximum item size for storages.

### CODE REVIEW

File: Content.Server/Storage/Components/ServerStorageComponent.cs [hunk 1]
```diff
@@ -134,7 +137,23 @@ public bool CanInsert(IEntity entity)
...
+
+            // Item is above max item size
+            if (_storageMaxItemSize != null && store?.Size > _storageMaxItemSize)
+            {
+                // Item is not whitelisted or there isn't a whitelist.
+                if (_whitelist == null || (_whitelist != null && !_whitelist.IsValid(entity)))
+                {
+                    return false;
+                }
+            }
```
Comment (PaulRitter) [inline Content.Server/Storage/Components/ServerStorageComponent.cs:154]: i dont get this part:
why not just
```suggestion
            // Item is not in whitelist
            if (_whitelist != null && !_whitelist.IsValid(entity))
            {
                return false;
            }

            // Item is above max item size
            if (_storageMaxItemSize != null && store?.Size > _storageMaxItemSize)
            {
                return false;
            }
```

**Review by PaulRitter — CHANGES_REQUESTED** [review]

File: Content.Server/Storage/Components/ServerStorageComponent.cs (hunk 1 above)
Comment (xTheLifex) [reply Content.Server/Storage/Components/ServerStorageComponent.cs:154]: Such solution would seem optimal at first, but it brings undesired behaviour.
It makes so that having a whitelist AND a max item size is pointless.

**Example:**
- Storage A has 40 capacity, and a max item size of 5.
- Item X occupies 10 of space.
- Item Y occupies 15 of space.
- Item Z occupies  5  of space.

Normally, Item X and Y do not fit into storage A due to its size.
However if a whitelist is applied containing the items we want to fit, the max size becomes pointless. And if the max size can dictate if something is rejected, if Item X is whitelisted, it will still not fit because of the max item size.

Instead, when using a whitelist AND a max item size, it is expected that all items that fit within the max item size to be allowed, and if a item does not match the max item size but is in the whitelist, it is also allowed.

The whitelist together with the max item size serves as a exception list to the max item size allowing further balancing options.

Comment (xTheLifex) [reply Content.Server/Storage/Components/ServerStorageComponent.cs:154]: > i dont get this part:
> why not just
> ```suggestion
>             // Item is not in whitelist
>             if (_whitelist != null && !_whitelist.IsValid(entity))
>             {
>                 return false;
>             }
> 
>             // Item is above max item size
>             if (_storageMaxItemSize != null && store?.Size > _storageMaxItemSize)
>             {
>                 return false;
>             }
> ```

Such solution would seem optimal at first, but it brings undesired behaviour.
It makes so that having a whitelist AND a max item size is pointless.

**Example:**
- Storage A has 40 capacity, and a max item size of 5.
- Item X occupies 10 of space.
- Item Y occupies 15 of space.
- Item Z occupies  5  of space.

Normally, Item X and Y do not fit into storage A due to its size.
However if a whitelist is applied containing the items we want to fit, the max size becomes pointless. And if the max size can dictate if something is rejected, if Item X is whitelisted, it will still not fit because of the max item size.

Instead, when using a whitelist AND a max item size, it is expected that all items that fit within the max item size to be allowed, and if a item does not match the max item size but is in the whitelist, it is also allowed.

The whitelist together with the max item size serves as a exception list to the max item size allowing further balancing options.

Comment (PaulRitter) [reply Content.Server/Storage/Components/ServerStorageComponent.cs:154]: what you are thinking off would be a "bypasslist" of sorts. thats not what the whitelist has been so far and i dont think we should change that. I'm not against adding a bypasslist tho

Comment (xTheLifex) [reply Content.Server/Storage/Components/ServerStorageComponent.cs:154]: The whitelist behaviour isn't changing if there is not a max item size, which is why its a nullable.

Comment (PaulRitter) [reply Content.Server/Storage/Components/ServerStorageComponent.cs:154]: > 
> 
> The whitelist behaviour isn't changing if there is not a max item size, which is why its a nullable.

but it will change behaviour if the max size is specified, hence my point.

i'll just say it bluntly, i wont merge it with the current behaviour. i'm not gonna block other maintainers from doing so, but i doubt other will have different opinions than mine.

Comment (metalgearsloth) [reply Content.Server/Storage/Components/ServerStorageComponent.cs:154]: I agree with Paul; every check should be passed or insertion should fail.

Comment (PJB3005) [reply Content.Server/Storage/Components/ServerStorageComponent.cs:154]: Conceptually, your proposal does an OR between the max_item_size check and whitelist check. So `fits = (max_item_size || whitelist)`. 

After a long and furious battle on Discord we came to the agreement that having a separate "bypass" list would be better. This is because an AND'd whitelist may still be useful *on top* of the proposed behavior (since it works with tags and components). 

So the logic would look like this: `fits = whitelist && (max_item_size || bypass_list)`. This would allow you to express "all tools below size 10 and also fits the jaws of life"

SS13 also has separate properties for this but the logic seems to make no sense to me (if `can_only_hold` then max item size is completely ignored, but `fits_ignoring_w_class` is union'd to `can_only_hold` which ???) similar: https://github.com/vgstation-coders/vgstation13/blob/2ded66406d928e1e2ff9f06a4253cd984ce1f00a/code/game/objects/items/weapons/storage/storage.dm#L13-L21

**Comment by ShadowCommander** [comment]
Are you going to come back to this?

**Comment by ShadowCommander** [comment]
Derelict for now. If someone wants to fix this up and continue it, feel free.
